platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/OutgoingVariantsMutationIntegrationTest.groovy

```diff
@@ -17,127 +17,183 @@
-                            }
+            def elements = configurations.create("elements") {
+                attributes.attribute(usage, "primary")
+                outgoing {
+                    variants {
+                        classes {
+                            attributes.attribute(usage, 'secondary')
+                            artifact(file('classes'))
                         }
                     }
                 }
             }
-            task mutateBeforeResolve {
-                doLast {
-                    def classes = configurations.compile.outgoing.variants['classes']
-                    classes.attributes.attribute(format, 'classes2')
+
+            elements.dependencies.addAllLater(provider {
+                // Try to mutate attributes late.
+                // We realize dependencies after we lock the configuration for mutation.
+                elements.outgoing.variants.classes {
+                    attributes.attribute(usage, 'tertiary')
                 }
-            }
-            task mutateAfterResolve {
+                []
+            })
+
+            task resolve {
+                def files = configurations.resolver.incoming.files
                 doLast {
-                    configurations.compile.resolve()
-                    def classes = configurations.compile.outgoing.variants['classes']
-                    classes.attributes.attribute(format, 'classes-dir')
+                    println(files*.name)
                 }
             }
         """
 
         when:
-        run 'mutateBeforeResolve'
+        fails("resolve")
 
         then:
-        noExceptionThrown()
+        failure.assertHasCause "Cannot change attributes of configuration ':elements' variant classes after it has been locked for mutation"
+    }
+
+    def "cannot declare capabilities after configuration is observed"() {
+        given:
+        buildFile << """
+            def elements = configurations.create("elements") {
+                attributes.attribute(usage, "primary")
+            }
+
+            elements.dependencies.addAllLater(provider {
+                // Try to add capabilities late.
+                // We realize dependencies after we lock the configuration for mutation.
+                elements.outgoing {
+                    capability("foo")
+                }
+                []
+            })
+
+            task resolve {
+                def files = configurations.resolver.incoming.files
+                doLast {
+                    println(files*.name)
+                }
+            }
+        """
 
         when:
-        fails("mutateAfterResolve")
+        fails("resolve")
 
         then:
-        failure.assertHasCause "Cannot change attributes of configuration ':compile' after it has been locked for mutation"
+        failure.assertHasCause("Cannot declare capability 'foo' on configuration ':elements' after it has been consumed as a variant.")
```

Comment:
💅 Why does the name of the test method "after configuration is Xed" and the part of the error message "after it has been Yed" have 2 different values here?  Shouldn't only the "Cannot Z on configuration" part change?  

Seems like it could be worth a tweak for more consistency here - these appear to be totally different problems, but they have the same root cause.

Reply:
I think this could use some improvement. The difference in messages comes from `FreezeableAttributeContainer` not accepting a reason for freezing, so it has to print a generic message. We could improve this but it touches a little more than I intended to here. This PR was intended for fixing bug not really for polishing error messages.